fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/ImplementInJoinRule.java

```diff
@@ -216,130 +220,210 @@ private ImmutableList<InSource> getInSourcesForRequestedOrdering(@Nonnull final
+                                                Objects.requireNonNull(explodeAliasToQuantifierMap.get(explodeAlias));
+                                        final var explodeExpression =
+                                                Objects.requireNonNull(quantifierToExplodeMap.get((Quantifier.ForEach)explodeQuantifier));
+
+                                        final var explodeValue = explodeExpression.getCollectionValue();
+
+                                        final var orderingResultsBuilder = ImmutableList.<OrderingPartWithSource>builder();
+                                        if (attemptedSortOrders == null) {
+                                            Verify.verify(!requestedOrdering.isExhaustive());
+                                            final InSource inSource =
+                                                    computeInSource(explodeValue, explodeQuantifier, null);
+                                            if (inSource != null) {
+                                                orderingResultsBuilder.add(new OrderingPartWithSource(null, inSource));
+                                            }
+                                        } else {
+                                            for (final var attemptedSortOrder : attemptedSortOrders) {
+                                                final InSource inSource =
+                                                        computeInSource(explodeValue, explodeQuantifier, attemptedSortOrder);
+
+                                                if (inSource != null) {
+                                                    orderingResultsBuilder.add(new OrderingPartWithSource(null, inSource));
+                                                }
+                                            }
+                                        }
+                                        return orderingResultsBuilder.build();
+                                    })
+                                    .collect(ImmutableList.toImmutableList());
+                    return Streams.stream(CrossProduct.crossProduct(suffix));
+                });
+    }
+
+    private static boolean isSupportedExplodeValue(@Nonnull final Value explodeValue) {
+        return explodeValue instanceof LiteralValue<?> ||
+                explodeValue instanceof QuantifiedObjectValue ||
+                explodeValue.isConstant();
+    }
+
+    @Nonnull
+    private static List<ProvidedSortOrder> attemptedProvidedSortOrdersForAny(final boolean isExhaustive) {
+        if (isExhaustive) {
+            // TODO consider creating the non-counterflow orders as well
+            return ImmutableList.of(ProvidedSortOrder.ASCENDING, ProvidedSortOrder.DESCENDING);
+        } else {
+            return ImmutableList.of(ProvidedSortOrder.ASCENDING);
+        }
+    }
+
+    @Nullable
+    @SuppressWarnings("unchecked")
+    private static InSource computeInSource(@Nonnull final Value explodeValue,
```

Comment:
This method was added since previously we had very similar code in two different places almost being just copy/pasta of one another.